### DISCUSSION TITLE
Purpose of finally

### DISCUSSION THREAD

**Discussion by mr1upmachine** [ReactiveX/rxjs#6191] [General]
What is the purpose of the `finally` operator function? These two code blocks seem identical to me:

```typescript
const someFunction = () => console.log('foo');
someObservable.pipe(
  tap({
    error: someFunction,
    complete: someFunction
  })
)
```

```typescript
someObservable.pipe(
  finally(() => console.log('foo'))
)
```

What purpose is `finally` serving other than being a more concise version of `tap`, if any? If that is all it is doing, then maybe we can add something to the documentation to reflect that? Or consider removing `finally` to make the API more concise? Or perhaps my personal understanding is lacking 😆

**Comment by josepot** [comment]
The 2 code snippets that you shared are not equivalent, because the callback passed to `finalize` also gets invoked when the subscriber unsubscribes from the observable (even if it hasn't completed/errored). For instance [this snipped](https://codesandbox.io/s/tap-vs-finalize-jj80y?file=/src/index.ts):

```ts
NEVER.pipe(
  tap({
    error: () => {
      console.log("tap got called!")
    },
    complete: () => {
      console.log("tap got called!")
    },
  }),
  finalize(() => {
    console.log("finalize got called!")
  }),
).subscribe().unsubscribe()
```

logs "finalize got called!" and it doesn't log "tap got called!".

**Reply by josepot** [reply]
Although, on that note: I wish that we could have an operator (or an overload of `finalize`) that would invoke a callback just when the unsubscription didn't happen because the source completed/errored... Something that behaved like this (the name is not great):

```ts
const onUnsubscribe = (callback: () => void) => <T>(source$: Observable<T>) =>
  new Observable<T>(observer => {
    let isSourceDone = false;

    const subscription = source$.subscribe(
      val => {
        observer.next(val);
      },
      e => {
        isSourceDone = true;
        observer.error(e);
      },
      () => {
        isSourceDone = true;
        observer.complete();
      }
    );

    return () => {
      if (isSourceDone) return;
      subscription.unsubscribe();
      callback();
    };
  });
```
